src/mlpack/methods/hdbscan/hdbscan_impl.hpp

```diff
@@ -0,0 +1,763 @@
+  {
+    // Initialize the elments
+    // There would be a max of 2*n-1 clusters
+    // formed after completion
+    // Initally all n points form a cluster
+    // After first merging there are
+    // (n-2) old clusters and 1 new cluster
+    // This goes on until at last there is one cluster
+    // (n-1) + n = (2n-1)
+    parentVector.set_size(2*n-1);
+    parentVector.fill(SIZE_MAX);
+    sizeVector.set_size(2*n-1);
+    for (size_t i = 0; i < n; i++)
+      sizeVector(i) = 1;
+    for (size_t i = n; i < 2*n-1; i++)
+      sizeVector(i) = 0;
+    nextLabel = n;
+  }
+  void UnionOfTwoClusters(size_t m, size_t n)
+  {
+    // Each union will form a new cluster
+    // with size as sum of its children
+    sizeVector(nextLabel) = sizeVector(m) + sizeVector(n);
+    parentVector(m) = nextLabel;
+    parentVector(n) = nextLabel;
+    nextLabel++;
+  }
+  size_t Find(size_t n)
+  {
+    // finds the parent
+    size_t root = n;
+    while (parentVector(root) != SIZE_MAX)
+      root = parentVector(root);
+    while (parentVector(n) != root)
+    {
+      size_t temp = parentVector(n);
+      parentVector(n) = root;
+      n = temp;
+      if (n == SIZE_MAX)
+        break;
+    }
+      return root;
+  }
+  size_t GetSize(size_t i)
+  {
+    return sizeVector(i);
+  }
+};
+
+/**
+ * Converts the input (sorted MST) to a single linkage tree.
+ * Each row of the output contaisn 3 values
+ * (cluster1 cluster2 d size)
+ *  d = distance between cluster1 and cluster2
+ *  size = size of cluster formed by merging cluster1 and cluster2
+ *  cluster1 and cluster2 may contain more than one point
+ *  Initially points are labelled as 0,1,2,...,(n-1)
+ *  and are single cluster(contain only one point)
+ *  Then we go on merging 2 clusters at a time until we
+ *  reach a point when all points are in one cluster
+ * 
+ *  This is done with help of unionForSingleLinkageTree
+ *  For every edge in MST, a merge operation is performed
+ *  which merges two clusters.
+ */
+
+template<typename NeighborSearch,
+         typename MetricType,
+         template<typename TreeMetricType,
+                  typename TreeStatType,
+                  typename TreeMatType> class TreeType >
+template<typename MatType>
+void HDBSCAN<NeighborSearch,
+             MetricType,
+             TreeType>::
+SingleLinkageTreeClustering(const MatType& inputMST,
+                          MatType& singleLinkageTree)
+{
+  singleLinkageTree.set_size(4, inputMST.n_cols);
+  size_t no = inputMST.n_cols + 1;
+  UnionForSingleLinkageClustering u(no);
+  size_t parentOfv1, parentOfv2;
+  double delta;
+  for (size_t i = 0; i < inputMST.n_cols; i++)
+  {
+    delta = inputMST(2, i);
+    parentOfv1 = u.Find(inputMST(0, i));
+    parentOfv2 = u.Find(inputMST(1, i));
+    singleLinkageTree.col(i) = arma::vec({(double)parentOfv1,
+      (double)parentOfv2, delta,
+      (double)(u.GetSize(parentOfv1) + u.GetSize(parentOfv2))});
+    u.UnionOfTwoClusters(parentOfv1, parentOfv2);
```

Comment:
This calculation does look pretty similar to the DualTreeBoruvka::EmitResults() calculation.  If you had access to the edge list produced by DualTreeBoruvka, would it make this code unnecessary?  I think (but I am not sure) that basically this code is the reverse of the code in EmitResults().  If that's true we can refactor DualTreeBoruvka so you get what you need out of it.  But I may have misunderstood, so please correct me if I am wrong.

Reply:
To construct SLT from MST, we go on merging the two closest clusters. 
For this to happen, the edges need to be sorted (w.r.t length), so I think EmitResults() is necessary . 
Though I think this part is not necessary 
```
// Make sure the edge list stores the smaller index first to
// make checking correctness easier
if (ind1 < ind2)
      {
        edges[i].Lesser() = ind1;
        edges[i].Greater() = ind2;
      }
      else
      {
        edges[i].Lesser() = ind2;
        edges[i].Greater() = ind1;
      }
```
If I have edge list and oldFromNew, this can be skipped (though I doubt if this will give any speed up).